plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/HudiSplitSource.java

```diff
@@ -0,0 +1,158 @@
+import static io.trino.plugin.hudi.HudiSessionProperties.isSizeBasedSplitWeightsEnabled;
+import static io.trino.plugin.hudi.HudiSessionProperties.shouldSkipMetaStoreForPartition;
+import static io.trino.plugin.hudi.query.HudiFileListerFactory.buildHudiFileLister;
+import static io.trino.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static java.util.concurrent.CompletableFuture.runAsync;
+import static java.util.stream.Collectors.toList;
+
+public class HudiSplitSource
+        implements ConnectorSplitSource
+{
+    private final AsyncQueue<ConnectorSplit> queue;
+
+    private TrinoException trinoException;
+
+    public HudiSplitSource(
+            ConnectorSession session,
+            HiveMetastore metastore,
+            Table table,
+            HudiTableHandle tableHandle,
+            FileSystem fs,
+            Map<String, HiveColumnHandle> partitionColumnHandleMap,
+            ExecutorService executor,
+            int maxSplitsPerSecond,
+            int maxOutstandingSplits)
+    {
+        boolean metadataEnabled = isHudiMetadataEnabled(session);
+        boolean shouldSkipMetastoreForPartition = shouldSkipMetaStoreForPartition(session);
+        HoodieTableMetaClient metaClient = buildTableMetaClient(fs.getConf(), tableHandle.getBasePath());
+        HoodieEngineContext engineContext = new HoodieLocalEngineContext(fs.getConf());
+        HoodieMetadataConfig metadataConfig = HoodieMetadataConfig.newBuilder()
+                .enable(metadataEnabled)
+                .build();
+        List<HiveColumnHandle> partitionColumnHandles = table.getPartitionColumns().stream()
+                .map(column -> partitionColumnHandleMap.get(column.getName())).collect(toList());
+
+        // TODO: fetch the query mode from config / query context
+        HudiFileLister hudiFileLister = buildHudiFileLister(
+                HudiQueryMode.READ_OPTIMIZED,
+                metadataConfig,
+                engineContext,
+                tableHandle,
+                metaClient,
+                metastore,
+                table,
+                partitionColumnHandles,
+                shouldSkipMetastoreForPartition);
+
+        this.queue = new ThrottledAsyncQueue<>(maxSplitsPerSecond, maxOutstandingSplits, executor);
+        HudiSplitBackgroundLoader splitLoader = new HudiSplitBackgroundLoader(
+                session,
+                fs,
+                tableHandle,
+                hudiFileLister,
+                queue,
+                executor,
+                createSplitWeightProvider(session));
+        runAsync(splitLoader, executor)
+                .exceptionally(throwable -> {
+                    trinoException = new TrinoException(GENERIC_INTERNAL_ERROR,
+                            "Failed to generate splits for " + table.getTableName(), throwable);
+                    queue.finish();
+                    return null;
+                });
```

Comment:
The fact that `queue` is partitally handled here and partially in `HudiSplitBackgroundLoader` makes it somewhat hard to read.
I would suggest to encapsulate all execution (including top level one) inside `HudiSplitBackgroundLoader`. 

Then here you would have sth like:
```
        HudiSplitBackgroundLoader splitLoader = new HudiSplitBackgroundLoader(
                session,
                fs,
                tableHandle,
                hudiFileLister,
                queue,
                e -> trinoException = new TrinoException(GENERIC_INTERNAL_ERROR,
                        "Failed to generate splits for " + table.getTableName(), e),
                executor,
                createSplitWeightProvider(session));
        splitLoader.start();
```

Note that I added one extra parameter to `HudiSplitBackgroundLoader` which is an exception listener (Consumer<Throwable>)